[WIP] add a remove subcommand to `ansible-galaxy collection`

s-hertel opened the pull request:
##### SUMMARY
Adds a 'remove' subcommand to `ansible-galaxy collection` to delete the directories of installed collections and dependencies

Currently this deletes matching collections from all search paths - should that be optional/not the default?

Fixes #67759

##### ISSUE TYPE
- Feature Pull Request

s-hertel committed:
Add a 'remove' option to 'ansible-galaxy collection' to remove collections and their dependencies

s-hertel committed:
add some initial tests

s-hertel committed:
Appease hacking script

Fix refactor

Pep8

ci_complete

s-hertel committed:
refactor to use the the collection's path instead of recreating it from the search path

ci_complete

jborean93 commented on lib/ansible/galaxy/collection.py:
```diff
@@ -584,6 +591,82 @@ def build_collection(collection_path, output_path, force):
     return collection_output
 
 
+def _collection_matches_requirement(collection, requirement, parent=None):
```
Is it worth having this in a separate function considering it's only used in 1 place?

jborean93 commented on lib/ansible/galaxy/collection.py:
```diff
@@ -584,6 +591,82 @@ def build_collection(collection_path, output_path, force):
+            namespace, name = collection_name.split('.')
+            collection_version = collection[1]
+
+            # do we want to remove all or first found? should it be a toggle?
```
One suggestion is that we only remove automatically if 1 was found but prompt the user if multiple were found. It would be something like

```
Found the following collections that match 'community.general'
    1 - /home/user/.ansible/collections/ansible_collections/community/general
    2 - /etc/ansible/collections/ansible_collections/community/general
    3 - /tmp/ansible_collections/community/general

Type the comma separate list of numbers to remove (a for all):
```

We could add an `--all` argument as well if we wanted a non-interactive run but we do run the risk of having an arg that most users want set by default.

s-hertel replied:
I implemented this to err on the side of caution as discussed. I used @webknjaz's suggestion of using the `--force` flag for non-interactive runs.

jborean93 commented on lib/ansible/galaxy/collection.py:
```diff
@@ -584,6 +591,82 @@ def build_collection(collection_path, output_path, force):
+                for local_collection in found_collections:
+                    local_collection.remove()
+
+                    # if an error is encountered above but --ignore_errors is set should dependencies be removed anyway instead?
```
No I don't think so, if a collection failed to be removed and had deps, then those deps should stay

jborean93 commented on lib/ansible/galaxy/collection.py:
```diff
@@ -584,6 +591,82 @@ def build_collection(collection_path, output_path, force):
+                    # if an error is encountered above but --ignore_errors is set should dependencies be removed anyway instead?
+                    if not no_deps:
+                        deps = [[identifier, requirement] for identifier, requirement in local_collection.dependencies.items()]
+                        remove_collections(deps, search_paths, ignore_errors, no_deps, parent=local_collection)
```
We have to be careful that we don't remove deps of another collection, i.e.

* community.general depends on ansible.posix
* other.collection also depends on ansible.posix

If we were to remove `community.general` then `ansible.posix` should not be removed because another collection relies on it.

s-hertel replied:
Good point!

s-hertel replied:
I've fixed this, let me know what you think.

jborean93 commented on lib/ansible/galaxy/collection.py:
```diff
@@ -615,7 +698,7 @@ def download_collections(collections, output_path, apis, validate_certs, no_deps
                 if requirement.api is None and requirement.b_path and os.path.isfile(requirement.b_path):
                     shutil.copy(requirement.b_path, to_bytes(dest_path, errors='surrogate_or_strict'))
                 elif requirement.api is None and requirement.b_path:
-                    temp_path = to_text(b_temp_path, errors='surrogate_or_string')
```
Does this cause a failure today?

s-hertel replied:
I'm not sure, I just noticed the typo and went looking to see if 'surrogate_or_string' was actually a thing.

s-hertel replied:
If you were still curious about this, looks like it can cause a failure
```
>>> from ansible.module_utils.common.text.converters import to_text
>>> to_text(b'[\xff]', 'ascii', errors='nope')
\Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/liveuser/ansible/lib/ansible/module_utils/common/text/converters.py", line 216, in to_text
    return obj.decode(encoding, errors)
LookupError: unknown error handler name 'nope'
```

jborean93 commented on lib/ansible/cli/galaxy.py:
```diff
@@ -1206,6 +1231,17 @@ def execute_remove(self):
+        requirements_file = context.CLIARGS['requirements']
+        no_deps = context.CLIARGS['no_deps']
+
+        requirements = self._require_one_of_collections_requirements(collections, requirements_file)['collections']
```
What's the behaviour when you have a requirements file of and a different version of that collection is installed

```yaml
collections:
- name: community.general
  version: 1.0.0
```

Looking through the code it seems like it will skip this collection due to [this check](https://github.com/ansible/ansible/pull/71800/files#diff-2de3c8cc3e7fae814d02b3058f0c8f9eR621). Is that the desired behaviour or should we be removing all the collections specified regardless of the version in a requirements file?

s-hertel replied:
That is intentional. Using that requirements file to remove a collection would be equivalent to `ansible-galaxy collection remove community.general:1.0.0` or `ansible-galaxy collection remove community.general:==1.0.0`; if a version other than 1.0.0 is installed, it won't be removed. Does that seem like surprising behavior?

jborean93 reviewed (commented):
Left a few comments, it seems like we have 2 choices when we want to implement the UI

* Make things completely headless and any user interaction is done through various arguments
* Simplify the CLI and prompt the user if any complex scenarios comes up, e.g. multiple collections found

It feels like we've been doing the 1st option a bit more in the other commands for `ansible-galaxy` but I'm wondering if that's ideal going forward. The main scenarios I think we need to deal with and decide how it will work are

* Multiple collections found
    * Will be prompted for confirmation of which one(s) to remove (or a for all)
    * Should it fail or require a further confirmation argument like `--force`
* Ignoring failures
    * Should we ignore if asked or consider a failure a failure
* Removing dependencies
    * Should we remove at all
    * Should we remove but consider an option to skip if desired
    * How do we deal with deps for other collections
    * Should we have a `clean` option to remove any orphaned deps

I'm not fundamentally against what you have right now in the PR I'm just concerned we will have an argument overflow and complicate the CLI options. The `install` command is arguably there and that's what we are trying to fix up for 2.11.

Personally I'm very much on the fence about `--ignore-errors`, if something fails then it should require manual intervention to fix up, just ignoring the error means the user will probably have to ignore the error in every future invocation. On the other hand I think `--no-deps` is fine, especially with it not being the default.

s-hertel commented:
Thanks for the feedback @jborean93.

In addition to your points, I was wondering what the behavior should be when something other than a collection name is provided (like a path to a tar.gz file). I think it would be really convenient if a requirements.yml file could be used for installing and removing collections. `verify` requires a collection name because the collection is compared to the one on a galaxy server, but I think in this case we could probably just use the collection source to find the collection name and then delete it if it's installed. What do you think? We could also maybe limit the scope of this and add support for a requirements.yml file later...

```
Multiple collections found
    Will be prompted for confirmation of which one(s) to remove (or a for all)
    Should it fail or require a further confirmation argument like --force
```
I really don't like interactive prompting. I toyed with an `--all` flag. I'm not really sure in what situations someone would *want* different versions of collections installed to multiple paths in the collections paths at the same time (since `ansible-galaxy install` tries not to do that it seems like it would have to be pretty intentional). I really like that `ansible-galaxy collection list` shows all paths a collection is in though, and it would be nice if removal worked in tandem.
```
Ignoring failures
    Should we ignore if asked or consider a failure a failure
```
I followed the --ignore-errors pattern from other options, but I'm also fine with removing it. The only case I'd be worried about is if a `requirements.yml` is supported, and there's no way to skip errors, I'd definitely want removing collections installed from tar.gz and git repo sources to also work.
```
...
    How do we deal with deps for other collections
```
I agree with your review comment. We should probably process everything up front and not remove deps that are being used by other collections. Should that be an error?
```
...
    Should we have a clean option to remove any orphaned deps
```
How would you know if it's an orphaned dep or standalone collection?

s-hertel committed:
Move tests to ansible-galaxy-collection

s-hertel committed:
Remove refactor and --ignore-errors

s-hertel committed:
Refactor to use the current dep resolver as the first step

Pass around CollectionRequirement objects for metadata instead of paths

Only remove unused dependencies

Prompt by default when multiple matches for a requirement are found

Be a little less verbose for requirements not found

Be more forgiving of different req types for requirements.yml compatibility

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -586,6 +594,180 @@ def build_collection(collection_path, output_path, force):
 
 
+def remove_collections(collections, search_paths, no_deps, force):
+    """Deletes the directories of installed collections that match the requirements list
```
Per PEP 257
```suggestion
    """Delete the directories of installed collections that match the requirements list.
```

s-hertel replied:
Thanks for these :-)

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -586,6 +594,180 @@ def build_collection(collection_path, output_path, force):
+
+
+def get_matching_collections(requirements, installed_collections, force):
+    """Gets the list of collections that match the requirements list
```
Also for consistency
```suggestion
    """Get the list of collections that match the requirements list.
```

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -586,6 +594,180 @@ def build_collection(collection_path, output_path, force):
+
+
+def filter_multiple_matches(req, collections, force):
+    """Filter multiple collections found for a requirement by prompting the user
```
```suggestion
    """Filter multiple collections found for a requirement by prompting the user.
```

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -586,6 +594,180 @@ def build_collection(collection_path, output_path, force):
+        b_req_name = to_bytes(req_name, errors='surrogate_or_strict')
+
+        if os.path.isfile(b_req_name) or urlparse(req_name).scheme.lower() in ['http', 'https'] or len(req_name.split('.')) != 2:
+            display.display("Skipping '%s' because it is not a valid collection name. The format namespace.name is expected." % req_name)
```
Shouldn't this explode loudly?

s-hertel replied:
Not in my opinion. I would like `ansible-galaxy collection remove` to work with a requirements.yml (using the same file to install, update, and remove collections is very convenient when there are a lot of them) and my `--ignore-errors` got a -1. Other suggestions of how to be usable with a requirements.yml are welcome.

webknjaz replied:
Well, then it needs a better logic to retrieve metadata from the concrete artifact pointers like URL/SCM (the tmp artifacts manager I'm working on should help with that). And also `len(req_name.split('.')) != 2` is a different class of an error. If an entry is clearly an invalid identifier, it shouldn't be silently ignored.

Also, if I'm reading this correctly, the message "skipping this" is printed but there's no actual skip (`continue`?) in code.

s-hertel replied:
I could download the tar.gz/git SCM, and parse the name from that to see if it's installed. If we do that, I think `verify` and upgrading a collection should probably work the same way.

Edit: Oops, responded before I refreshed to see your response above.

Btw- `len(req_name.split('.')) != 2` is less intended to validate the collection name, than identify things that are not collection names. Checking it's a valid collection identifier is not what this was intended to do, though we could add that (but this didn't really seem like the right place - better in the dep resolver maybe).

webknjaz replied:
Yes, since I'm working on a standardized way to do this already, you probably shouldn't spend effort on implementing this.

s-hertel replied:
:100:

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -586,6 +594,180 @@ def build_collection(collection_path, output_path, force):
+        elif result in ['n', 'none']:
+            indices = []
+            break
+        else:
+            try:
+                indices = [int(r) if int(r) in indices else int('invalid') for r in result.split(',')]
+            except ValueError:
+                pass
+            else:
+                break
```
Nesting is unnecessary here
```suggestion
        try:
            indices = [int(r) if int(r) in indices else int('invalid') for r in result.split(',')]
        except ValueError:
            pass
        else:
            break
```

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -586,6 +594,180 @@ def build_collection(collection_path, output_path, force):
+            break
+        else:
+            try:
+                indices = [int(r) if int(r) in indices else int('invalid') for r in result.split(',')]
```
I'm puzzled by `int('invalid')`: what is it?
```python
In [1]: int('invalid')                                                        
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-4-adc72949be0b> in <module>
----> 1 int('invalid')

ValueError: invalid literal for int() with base 10: 'invalid'
```

webknjaz replied:
Oh... You're triggering `ValueError` this way. This may need a code comment to reduce the confusion.

s-hertel committed:
Apply suggestions from code review

Co-authored-by: Sviatoslav Sydorenko <wk.cvs.github@sydorenko.org.ua>

s-hertel committed:
Update lib/ansible/galaxy/collection/__init__.py

Co-authored-by: Sviatoslav Sydorenko <wk.cvs.github@sydorenko.org.ua>

s-hertel committed:
Add a comment

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -586,6 +594,180 @@ def build_collection(collection_path, output_path, force):
+
+
+def get_dependencies(collections, installed_collections):
+    """Finds all dependencies of the collections list in the list of all collections
```
```suggestion
    """Find all dependencies of the collections list in the list of all collections.
```

webknjaz commented on lib/ansible/galaxy/collection/__init__.py:
```diff
@@ -586,6 +594,180 @@ def build_collection(collection_path, output_path, force):
+
+
+def get_parents(dependencies, skip_parents, installed_collections):
+    """Finds all collections relying on a list of dependencies
```
```suggestion
    """Find all collections relying on a list of dependencies.
```

s-hertel committed:
moar

Co-authored-by: Sviatoslav Sydorenko <wk.cvs.github@sydorenko.org.ua>

s-hertel committed:
Fix skipping and add TODOs

s-hertel commented on test/integration/targets/ansible-galaxy-collection/tasks/remove.yml:
```diff
@@ -0,0 +1,121 @@
+      - '"namespace1.collection1 (2.0.0) was removed successfully" in result.stdout_lines'
+      - '"namespace1.collection2 (2.0.0) was removed successfully" in result.stdout_lines'
+
+- name: test moving the same collection
```
```suggestion
- name: test removing the same collection
```

s-hertel committed:
typo

reinvantveer commented:
This feature would be great: I'm trying to prune unused collections in order to reduce a Docker image built on Ansible. A lot of data resides in the ansible_collections folder and removing using ansible-galaxy would be a very useful feature. Can I do something to help?

webknjaz commented:
> Can I do something to help?

This needs to be rebased/remade now that we've replaced the dependency resolver completely so it'll take some time to get done. We still have some time before the feature freeze, though.

reinvantveer commented:
OK I'll take a look this evening, see if I can contribute or if it's beyond my capabilities

webknjaz commented on lib/ansible/cli/galaxy.py:
```diff
@@ -239,10 +241,21 @@ def add_init_options(self, parser, parents=None):
-        remove_parser = parser.add_parser('remove', parents=parents, help='Delete roles from roles_path.')
-        remove_parser.set_defaults(func=self.execute_remove)
+        if parser.metavar == 'COLLECTION_ACTION':
+            remove_parser = parser.add_parser('remove', parents=parents, help='Delete collections installed in the collections path')
```
@s-hertel did we talk about `remove` vs `uninstall`? It occurred to me that people switching from pip may expect `uninstall`.

reinvantveer replied:
To be honest this was the first thing I tried before finding out that this wasn't something in ansible-galaxy yet

reinvantveer commented:
I think it's clear enough for me to wrap my head around. The question is from where I can take things to contribute. I can start by forking & branching ansible:devel and see if I can reconcile the commits by @s-hertel with the changes in the code base. I could open a new PR then, referencing this on. Would this be workable?

s-hertel commented:
@reinvantveer Thanks for the interest! This needs to be reworked from scratch to fit naturally with the changes made in 595413d11346b6f26bb3d9df2d8e05f2747508a3. Trying to salvage the commits here is not worth the effort. Finishing `--upgrade`/explicit upgrades is the immediate priority (https://github.com/ansible/ansible/pull/73336) and then I plan to open a new PR to address #67759.

webknjaz commented:
@reinvantveer I just wanted to add that this improvement is on our roadmap as a part of our effort to improve the `ansible-galaxy collection` subcommands UX for ansible-core v2.12 (https://docs.ansible.com/ansible/devel/roadmap/ROADMAP_2_11.html). That PR Sloane mentioned was a colossal refactoring replacing the old in-house way of resolving collection dependencies.

That said if you want to cheap in, feel free to go ahead and submit a PR. Yes, you'd need to branch out of `ansible:devel`, and maybe the tests added in this PR would be useful, so you may want to add Sloane's fork as a Git remote and cherry-pick those. Keep in mind, however, that we want this done before the feature freeze on Feb 12 so we may have to hake over and extend/polish your work at some point.

You may open a pull request marked as a `Draft` and get regular feedback while working on it, this will help us see how ready the patch is and help you out if you have any questions.

P.S. If you feel like it's too much, just make sure to tell us to pick it up before it's too late. Have fun :)
P.P.S. Oh, and you may look at https://docs.ansible.com/ansible/devel/community/ / https://docs.ansible.com/ansible/devel/community/development_process.html#making-your-pr-merge-worthy for some contributing guidance + there's an `#ansible-devel` channel on Freenode IRC, come by to chat if you need any help.

reinvantveer commented:
OK, Feb 12 I'd certainly not make that, but thanks!

s-hertel commented:
@reinvantveer (/anyone else following this PR) - Just to clarify, it was discussed as a feature for 2.11 and would have needed to be done by the Feb 12th deadline, but it was not an official part of the roadmap linked above (more of a stretch goal). We've decided against trying to rush it into 2.11 at this point so it can be more thoughtfully planned out, but I've created another draft on top of the dep resolver changes here: #73464.
